### ISSUE TITLE
add explicit language switcher

### ISSUE THREAD

**Issue by jeanpommier** [georchestra/georchestra#3314]
Hi,
While working on https://geopresovregion.sk/geonetwork geOrchestra instance, we hacked a language switcher functionality. Because when sharing screeen between French and Slovak people, it's often useful to be able to all switch to a common eng language ;o)
As I said, this was a hack (in the header, but you have to manage things differently depending on whether it's a GN url or not). We'd like to contribute this functionnality back to upstream georchestra project. But for now, we're not really clear about how to best implement it.
I'm thinking of coding the logic in the security-proxy, since this is the centerpiece of the platform: it intercepts all requests, so this seems to me like the best place to articulate the language management.
What do you think ?

**Comment by RemiDesgrange** [comment]
I think that the default should be the app sending you data based on the language of your browser, with a fallback on english (we don't have any translation of georchestra for bzh, which is a shame, I agree). Your use case is interesting here is workflow:

* the user view a page on the georchestra plateform, the app sends content based on his browser Accept-Language
* user switch language with the switch. Then a cookie is persisted or something similar, that makes the SP change the `Accept-Language`

WDYT ?

**Comment by jeanpommier** [comment]
Yes, sure, the default should be your language as detected by the browser. 
Your proposal looks quite neat and simple to implement, I like it. Thanks

**Comment by fvanderbiest** [comment]
I was not sure Geonetwork respects the `Accept-Language` header but it seems he does !

**Comment by fgravin** [comment]
Hi, here a proposal to manage the language in a cookie.
Tell me what do you think

In header.jsp

- Get the cookie

```java
if(forcedLang == null) {
    Map<String, Cookie> cookieMap = new HashMap<>();
    if (request.getCookies() != null) {
        for (Cookie cookie : request.getCookies()) {
            cookieMap.put(cookie.getName(), cookie);
        }
    }
    if (cookieMap.get("geor-lang") != null) {
        forcedLang = cookieMap.get("geor-lang").getValue();
    }
}

```

- Set the cookie (reload the application)


```jsp
<div class="lang-switcher">
    <a onclick="return georSwitchLang('en')" class='<%=lang.equals("en") ? "active" : ""%>' href="">EN</a> |
    <a onclick="return georSwitchLang('fr')" class='<%=lang.equals("fr") ? "active" : ""%>' href="">FR</a>
</div>
```
```js
function georSwitchLang(lang) {
  document.cookie = "geor-lang=" + lang + ";path=/";
  window.top.location.reload();
  return false;
}
```

Example in Geonetwork

- Read from the cookie

```js
try {
  iso2lang = document.cookie.split('; ')
    .find(row => row.startsWith('geor-lang='))
    .split('=')[1];
} catch (e) {}
```

- Set the cookie if language is forced with a url param, from CMS for instance (reload the header, remove the param)
```js
const urlParams = new URLSearchParams(window.location.search);
const forcedLang = urlParams.get('georlang');
if(forcedLang) {
    document.cookie = "geor-lang=" + forcedLang + ";path=/";
    document.getElementById('georchestra-header').contentWindow.location.reload();
    window.history.replaceState({}, document.title, window.location.href.replace('georlang=' + forcedLang,''))
}
```

**Comment by fvanderbiest** [comment]
Thanks for taking the time to elaborate a proposal Florent.

I have to admit I'm not a big fan of having a GET lang parameter... 
Why not sticking to what @RemiDesgrange proposed above with a local (session) change of the `Accept-Language` header between the SP and apps ?

**Comment by fgravin** [comment]
Thanks for the feedback.

In this proposal, the header takes the browser language by default, and provides a language switcher to change the language.
This results in a language cookie which is stored on the browser and which is used by geonetwork 3 UI and mapstore UI to load the correct language. So there are no url parameters involved here.

The url parameter is just here to force the language if you come from outside georchestra and ask for a specific language, eg you come from a CMS which has its own language switcher, which is not behind the security-proxy and explicitly asks to load mapstore in french while the browser is in english. It's easier to provide a force-language url parameter, than ask to CMS administrator to insert javascript to store the cookie (could be discussed). Note that the parameter is removed from the url when the cookie is set, not to disturb the mainstream behavior.

For the `Accept-Language` header, I think it could be useful but in this specific usecase, i'm not sure (maybe for backend served UI only). To me, this header is attached to a request (eg front to api request) but can't be used by static frontend resources.
For instance, geonetwork 3 UI detects the language from the UI (javascript) and uses this language to add the correct `Accept-Language` header to all requests that the frontend performs on the backend (with a request interceptor). Mapstore language detection is also managed in the javascript, where you don't have the request header information.
Overall, the most important is that all the application frontend (static HTML resources) are in the correct language, which may not be handled by the request header (as far as I understand).

Let me know if you have information that i'm missing here.
Thanks
